src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/Maintainability/CodeMetricsAnalyzerTests.cs

```diff
@@ -102,14 +105,347 @@ End Class
 # FORMAT:
 # 'RuleId'(Optional 'SymbolKind'): 'Threshold'
 
-CA1501: 1
+CA1501: 0
 ";
             DiagnosticResult[] expected = new[] {
-                // Test0.vb(5, 7): warning CA1501: 'BaseClass' has an object hierarchy '2' levels deep within the defining module. If possible, eliminate base classes within the hierarchy to decrease its hierarchy level below '2': 'BaseClass, Object'
-                GetBasicCA1501ExpectedDiagnostic(5, 7, "FirstDerivedClass", 2, 2, "BaseClass, Object")};
+                GetBasicCA1501ExpectedDiagnostic(5, 7, "FirstDerivedClass", 1, 1, "BaseClass")};
             await VerifyBasicAsync(source, additionalText, expected);
         }
 
+        [Fact, WorkItem(1839, "https://github.com/dotnet/roslyn-analyzers/issues/1839")]
+        public async Task CA1501_ExcludesTypesInSystemNamespaceByDefault()
+        {
+            await new VerifyCS.Test
+            {
+                TestCode = "public class MyUC : System.Windows.Forms.UserControl {}",
+                ReferenceAssemblies = AdditionalMetadataReferences.DefaultWithWinForms,
+            }.RunAsync();
+
+            await new VerifyVB.Test
+            {
+                TestCode = @"
+Public Class MyUC
+    Inherits System.Windows.Forms.UserControl
+End Class",
+                ReferenceAssemblies = AdditionalMetadataReferences.DefaultWithWinForms,
+            }.RunAsync();
+        }
+
+        [Fact, WorkItem(1839, "https://github.com/dotnet/roslyn-analyzers/issues/1839")]
+        public async Task CA1501_UserConfigurationOverridesDefaultValue()
+        {
+            var editorConfigText = "dotnet_code_quality.CA1501.inheritance_excluded_type_names = T:SomeClass";
+
+            await new VerifyCS.Test
+            {
+                TestState =
+                {
+                    Sources = { "public class MyUC : System.Windows.Forms.UserControl {}" },
+                    AdditionalFiles = { (".editorconfig", editorConfigText) },
+                    ExpectedDiagnostics = { GetCSharpCA1501ExpectedDiagnostic(1, 14, "MyUC", 7, 6, "UserControl, ContainerControl, ScrollableControl, Control, Component, MarshalByRefObject, Object") },
```

Comment:
I think we probably want `additional_inheritance_excluded_symbol_names`:

1. Users can exclude even namespaces, not just types
2. I don't think users would want to remove the exclusion for `System.*` when adding a new value. We should probably just assume `System.*` is always excluded.

Reply:
> I don't think users would want to remove the exclusion for System.* when adding a new value. We should probably just assume System.* is always excluded.

I was thinking that people at Microsoft might be interested. Assuming they are using this analyzer.

Reply:
I have addressed point 1 let me know if you think that `System.*` exclusion shall be always activated.

Reply:
I would think the number of external people who will accidentally get affected when they configure a new exclusion would be much broader then those working in System.* namespace. We can possibly add a separate option to remove the System.* exclusion.

Reply:
Fine by me, just as a reminder, with this change we will cause a breaking change.

Reply:
Done